File: S02/week14/js/InsertionSortList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// Runtime 18 ms Beats 58.65% of users with Java
// Memory 44.00 MB Beats 65.77% of users with Java
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode sorted = head;
        head = head.next;
        
        sorted.next = null;
        while(head != null) {
            ListNode cur = sorted;
            ListNode prev = null;
            ListNode target = head;
            
            head = head.next;
            while (cur != null) {
                if (cur.val < target.val) {
                    if (cur.next == null) {
                        cur.next = target;
                        target.next = null;
                        break;
                    } else {
                        prev = cur;
                        cur = cur.next;
                    }
                } else {
                    target.next = cur;
                    if (prev == null) {
                        sorted = target;
                    } else {
                        prev.next = target;
                    }

                    break;
                }
            }
        }

        return sorted;
    }
}

File: week14/seokjoon/InsertionSortList.kt
//200ms, 37.08mb
class Solution {
    fun insertionSortList(head: ListNode?): ListNode? {
        var curNode = head
        val newHeadNode = ListNode(-5001)
        while(curNode != null){
            var curNewNode = newHeadNode
            while(curNewNode.next != null && curNewNode.next!!.`val` < curNode!!.`val` ){  // curNewNode의 `val`이 삽입할 Node의 `val`보다 크거나 null일 때까지 curNewNode 위치를 옮겨준다.
                curNewNode = curNewNode.next
            }
            if(curNewNode.next == null){ //curNewNode.next가 null이면 next에 새로운 Node 추가
                curNewNode.next = ListNode(curNode!!.`val`)
            }else { //curNewNode.next가 null이 아니면 next 사이에 새로운 Node 추가 후 이어붙여준다. 
                curNewNode.next = ListNode(curNode!!.`val`).apply {
                    next = curNewNode.next
                }
            }
            curNode = curNode?.next
        }
        return newHeadNode?.next
    }
}
